lib/templates.php

```diff
@@ -0,0 +1,74 @@
+<?php
+/**
+ * Register Gutenberg core block editor templates.
+ *
+ * @package gutenberg
+ */
+
+/**
+ * Registers Gutenberg core block editor templates.
+ */
+function gutenberg_register_templates() {
+	if (
+		get_option( 'gutenberg-experiments' ) &&
+		! array_key_exists( 'gutenberg-full-site-editing', get_option( 'gutenberg-experiments' ) )
+	) {
+		return;
+	}
+
+	register_post_type(
+		'wp_template',
+		array(
+			'labels'       => array(
+				'name' => __( 'Templates', 'gutenberg' ),
+			),
+			'show_in_rest' => true,
+		)
+	);
+
+	$template_query = new WP_Query(
+		array(
+			'post_type' => 'wp_template',
+			'name'      => 'single-post',
+		)
+	);
+	if ( ! $template_query->have_posts() ) {
+		wp_insert_post(
+			array(
+				'post_type'    => 'wp_template',
+				'post_name'    => 'single-post',
+				'post_content' => '<!-- wp:post-title /--><!-- wp:post-content /-->',
+			)
+		);
+	}
+}
+add_action( 'init', 'gutenberg_register_templates' );
+
+/**
+ * Adds the relevant template ID to the block editor settings.
+ *
+ * @param array $settings The editor settings.
+ *
+ * @return array The editor settings with the relevant template ID.
+ */
+function gutenberg_templates_filter_block_editor_settings( $settings ) {
+	if (
+		get_option( 'gutenberg-experiments' ) &&
+		! array_key_exists( 'gutenberg-full-site-editing', get_option( 'gutenberg-experiments' ) )
+	) {
+		return $settings;
+	}
+
+	$template_query = new WP_Query(
+		array(
+			'post_type' => 'wp_template',
+			'name'      => 'single-post',
```

Comment:
Naturally, this logic will need to be fleshed out with looking up the specific template for a given post. For example, if editing a `movie` post type, it should look for `single-movie` before `single`. And if `single` does not exist, then `singular`, and so on.

Reply:
Yes, see #4659